apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/SessionPoolTests.cs

```diff
@@ -0,0 +1,443 @@
+                    .ConfigureAwait(false);
+
+                Assert.NotSame(session, session2);
+                Assert.Equal(1, _createdSessions.Count);
+                Assert.Equal(2, _createdSessions[client].Count);
+            }
+        }
+
+        [Fact]
+        public async Task MaxPoolSizeNotViolated()
+        {
+            using (var pool = new SessionPool())
+            {
+                pool.Options.MaximumPooledSessions = 2;
+
+                var client = CreateMockClient();
+                var sessions = await Task.WhenAll(
+                        pool.CreateSessionFromPoolAsync(
+                            client.Object, s_defaultName.ProjectId,
+                            s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None),
+                        pool.CreateSessionFromPoolAsync(
+                            client.Object, s_defaultName.ProjectId,
+                            s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None),
+                        pool.CreateSessionFromPoolAsync(
+                            client.Object, s_defaultName.ProjectId,
+                            s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None))
+                    .ConfigureAwait(false);
+
+                pool.ReleaseToPool(client.Object, sessions[0]);
+                pool.ReleaseToPool(client.Object, sessions[1]);
+                pool.ReleaseToPool(client.Object, sessions[2]);
+
+                Assert.Equal(
+                    2, pool.GetPoolSize(
+                        client.Object, s_defaultName.ProjectId,
+                        s_defaultName.InstanceId, s_defaultName.DatabaseId));
+            }
+        }
+
+        [Fact]
+        public async Task MaxActiveViolationThrows()
+        {
+            using (var pool = new SessionPool())
+            {
+                pool.Options.WaitOnResourcesExhausted = false;
+                pool.Options.MaximumActiveSessions = 2;
+                bool exceptionThrown = false;
+
+                var client = CreateMockClient();
+                await pool.CreateSessionFromPoolAsync(
+                        client.Object, s_defaultName.ProjectId,
+                        s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+                await pool.CreateSessionFromPoolAsync(
+                        client.Object, s_defaultName.ProjectId,
+                        s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                    .ConfigureAwait(false);
+                try
+                {
+                    await pool.CreateSessionFromPoolAsync(
+                        client.Object, s_defaultName.ProjectId,
+                        s_defaultName.InstanceId, s_defaultName.DatabaseId, null, CancellationToken.None)
+                        .ConfigureAwait(false);
```

Comment:
thats true, but I'm leaning towards keeping it the way it is as its self documenting that an exception is expected versus "what is that assert(false) doing there?".  Then you'll have to add a comment to explain it, etc etc.